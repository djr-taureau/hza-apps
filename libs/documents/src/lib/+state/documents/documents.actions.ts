import { createAction, props } from '@ngrx/store';
import { Update } from '@ngrx/entity';
import { Document } from './../../models/document.model';
import { CodeTable } from '../../models/code-table.model';



export const loadDocs = createAction('[documents] LOAD_DOCS');

export const loadDoc = createAction('[documents] Load Document', props<{ document: Document }>());
export const selectDoc = createAction('[documents] Select Document', props<{ ID: number }>());

export const loadDocTypes = createAction('[documents] LOAD_DOCS_TYPES');
export const loadDocTypesSuccess = createAction('[documents] LOAD_DOCTYPES_SUCCESS', props<{ docTypes: CodeTable[] }>())

export const loadDocsSuccess = createAction('[documents] LOAD_DOCS_SUCCESS', props<{ documents: Document[] }>());
export const loadDocsFail = createAction('[documents] LOAD_DOCS_FAIL', props<{ error: Error }>());
