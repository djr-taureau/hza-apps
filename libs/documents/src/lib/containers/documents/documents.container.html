<div class="hza-elevatation container">
  <!-- <hza-section>
    <hza-control-bar>
      <h2 class="title-bar" hzaSectionTitle>Repository</h2>
    </hza-control-bar>
  </hza-section> -->
  <!-- <hza-section>
    PLACEHOLDER FOR DOC UPLOAD
  </hza-section> -->
  <hza-section>
    <hza-control-bar>
      <h2 class="title-bar" hzaSectionTitle>Documents</h2>
    </hza-control-bar>
  </hza-section>
  <hza-doc-list [documents]="documents$ | async"></hza-doc-list>
</div>
