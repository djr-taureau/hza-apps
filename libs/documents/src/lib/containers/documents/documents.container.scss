:host {
  .container {
    width: 100%;
    padding: 0;
    margin: 0;
    background-color: rgba(255, 255, 255, 0.7);
    border: 2px solid #37678e;
    border-radius: 5px;
    box-shadow: #000 5px 5px 10px;
    color: #000;
    float: left;
    font-family: 'Segoe UI', Verdana, Helvetica, sans-serif;
    line-height: 24px;
    margin: 0.5em 0 0 0.5em;
    z-index: 1;
  }
}

.title-bar {
  font-size: 1.5em;
  font-weight: 600;
}

hza-control-bar {
  height: 2em;
  z-index: 5;
}

.mat-flat-button {
  background-color: #d3dce0;
  border: 1px solid #787878;
  border-radius: 5px;
  height: 1.5em;
  color: #000000;
  cursor: pointer;
  font-family: Arial;
  font-size: 1em;
  font-weight: 700;
  margin-right: 0.25em;
  padding: 0 0.5em;
  text-align: center;
  // vertical-align: middle;
  width: auto;

  &:disabled {
    background-color: #d3d3d3;
    color: #808080;
    cursor: default;
  }
}

.mat-table {
  flex: 1;
}

select {
  background-color: #ADD8E6;
  color: #333333;
  cursor: default;
  font-family: Arial;
  font-size: 1em;
  height: 1.75em;
  margin: .25em;
  max-width: 28em;
  padding: 0 .25em;
  vertical-align: middle;
}
