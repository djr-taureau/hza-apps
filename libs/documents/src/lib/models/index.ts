// start:ng42.barrel
export * from './code-table.model';
export * from './doc-detail-form.config';
export * from './document.model';
// end:ng42.barrel

