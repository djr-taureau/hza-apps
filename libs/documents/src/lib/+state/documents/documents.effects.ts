import { Injectable } from '@angular/core';
import { Actions, Effect, ofType, createEffect } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { DocumentsService } from './../../services/documents.service';
import * as DocActions from './documents.actions';
import { DocsFacade } from './documents.facade';

@Injectable()
export class DocEffects {
  // ** the use of switchMap is important here because as soon as it completed task it turns the
  // ** the observable off and waits for the next request or action to be dispatched
  loadDocs$ = createEffect(() =>
    this.actions$.pipe(
      ofType(DocActions.loadDocs),
      switchMap(() =>
        this.docService.getDocuments().pipe(
          map(documents => DocActions.loadDocsSuccess({ documents })),
          catchError(error => of(DocActions.loadDocsFail(error))),
        ),
      ),
    ),
  );
  
    loadDocTypes$ = createEffect(() =>
    this.actions$.pipe(
      ofType(DocActions.loadDocTypes),
      switchMap(() =>
        this.docService.getDocumentTypes().pipe(
          map(docTypes => DocActions.loadDocTypesSuccess({ docTypes }))
        ),
      ),
    ),
  );



  constructor(private actions$: Actions, private docService: DocumentsService, private facade: DocsFacade) {}
}
