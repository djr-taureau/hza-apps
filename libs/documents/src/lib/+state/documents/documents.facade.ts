import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';

import { State, docsQuery } from '.';
import * as DocActions from './documents.actions';
import { Observable } from 'rxjs';

// ** The use of facades is good with managing state
// ** because it avoids having to inject your store into every container that needs to interact
// **
@Injectable()
export class DocsFacade {
  documents$ = this.store.select(docsQuery.selectAllDocs);
  docsLoaded$ = this.store.select(docsQuery.loadedDocs);

  docTotal$ = this.store.select(docsQuery.selectDocTotal);
  selectedDoc$ = this.store.select(docsQuery.selectCurrentDoc);

  // ** The Store gets injected one time for each facade

  constructor(private store: Store<State>) {}

  loadDocs() {
    this.store.dispatch(DocActions.loadDocs());
  }
  
  loadDocTypes() {
    this.store.dispatch(DocActions.loadDocTypes());
  }

  selectDoc(ID: number) {
    this.store.dispatch(DocActions.selectDoc({ ID }));
  }
}
