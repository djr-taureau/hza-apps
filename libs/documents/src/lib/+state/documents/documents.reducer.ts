import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { Action, createReducer, on } from '@ngrx/store';
import { Document } from './../../models/document.model';
import * as DocActions from './documents.actions';

// ** This reducer is what each reducer will look like for an entity with the standard
// ** state interface
export const docsFeatureKey = 'documents';

export interface DocsState extends EntityState<Document> {
  selectedDocId: number | null;
  isLoading: boolean;
  loaded: boolean;
}

export const adapter: EntityAdapter<Document> = createEntityAdapter<Document>({
  selectId: (document: Document) => document.id,
  sortComparer: false,
});

export const docsInitialState: DocsState = adapter.getInitialState({
  ids: [],
  selectedDocId: null,
  isLoading: false,
  loaded: false,
});

export const reducer = createReducer(
  docsInitialState,
  /**
   * The addMany function provided by the created adapter
   * adds many records to the entity dictionary
   * and returns a new state including those records. If
   * the collection is to be sorted, the adapter will
   * sort each record upon entry into the sorted array.
   */
  //  ** when this action is dispatched isLoading set to true
  on(DocActions.loadDocs, state => ({
    ...state,
    isLoading: true,
  })),
  on(DocActions.loadDocsSuccess, (state, { documents }) => adapter.addMany(documents, state)),
  // ** When the documents have loaded successfully, isLoaded toggled off again
  // ** loaded set to TRUE
  on(DocActions.loadDocsSuccess, state => ({
    ...state,
    isLoading: false,
    loaded: true,
  })),
  /**
   * The addOne function provided by the created adapter
   * adds one record to the entity dictionary
   * and returns a new state including that records if it doesn't
   * exist already. If the collection is to be sorted, the adapter will
   * insert the new record into the sorted array.
   */
  // ** Select an individual document for docment detail
  on(DocActions.loadDoc, (state, { document }) => adapter.addOne(document, state)),
  on(DocActions.selectDoc, (state, { id }) => ({
    ...state,
    selectedDocId: id,
  })),
);

export function docsReducer(state: DocsState | undefined, action: Action): DocsState {
  return reducer(state, action);
}

// ** selectors are slices of your state that can be rolled up with other reducers
// **

export const getSelectedDocId = (state: DocsState) => state.selectedDocId;
export const isLoading = (state: DocsState) => state.isLoading;
export const loaded = (state: DocsState) => state.loaded;


// get the selectors
const { selectIds, selectEntities, selectAll, selectTotal } = adapter.getSelectors();

export const selectDocIds = selectIds;

export const selectDocEntities = selectEntities;

export const selectAllDocs = selectAll;

export const selectDocTotal = selectTotal;
