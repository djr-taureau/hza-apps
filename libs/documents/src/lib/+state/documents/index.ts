import { createSelector, createFeatureSelector, combineReducers, Action, ActionReducerMap } from '@ngrx/store';
import * as fromDocs from './documents.reducer';


// ** This is an example of rolling up your reducers for a feature
// ** With this you will then include this in your feature module
export const docRepoFeatureKey = 'documents';


export interface State {
	[docRepoFeatureKey]: fromDocs.DocsState;
}

/** Provide reducer in AoT-compilation happy way */
export const reducers: ActionReducerMap<State> = {
	documents: fromDocs.docsReducer
}

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const selectDocRepoState = createFeatureSelector<fromDocs.DocsState>(docRepoFeatureKey);

// ** SELECTORS FOR LOANS


export const selectDocIds = createSelector(
  selectDocRepoState,
  fromDocs.selectDocIds
);
export const selectDocEntities = createSelector(
  selectDocRepoState,
  fromDocs.selectDocEntities
);
export const selectAllDocs = createSelector(
  selectDocRepoState,
  fromDocs.selectAllDocs
);
export const selectDocTotal = createSelector(
  selectDocRepoState,
  fromDocs.selectDocTotal
);
export const selectCurrentDocId = createSelector(
  selectDocRepoState,
  fromDocs.getSelectedDocId
);

export const selectDocTypes = createSelector(
  selectDocRepoState,
  fromDocs.getDocTypes
);

export const isLoadingDocs = createSelector(
  selectDocRepoState,
  fromDocs.isLoading
);

export const loadedDocs = createSelector(
  selectDocRepoState,
  fromDocs.loaded
);
 
export const selectCurrentDoc = createSelector(
  selectDocEntities,
  selectCurrentDocId,
  (docEntities, docId) => docEntities[docId]
);



export const docsQuery = {
	selectDocIds,
	selectDocEntities,
	selectAllDocs,
	selectDocTotal,
	isLoadingDocs,
	loadedDocs,
	selectCurrentDoc,
  selectDocTypes
};
