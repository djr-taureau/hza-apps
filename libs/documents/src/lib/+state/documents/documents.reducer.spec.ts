import * as DocActions from './documents.actions';
import * as fromDocs from './documents.reducer';
import { mockDoc1, mockDoc2, Document } from '../../models';
import { mockDocType1, mockDocType2, CodeTable } from '../../models/code-table.model';

describe('DocumentsReducer', () => {
	const doc1 = mockDoc1 as Document;
	const doc2 = mockDoc2 as Document;
	const docType1 = mockDocType1 as CodeTable;
	const docType2 = mockDocType2 as CodeTable;

	const docsInitialState: fromDocs.DocsState = {
		ids: [ doc1.ID, doc2.ID ],
		entities: {
			[doc1.ID]: doc1,
			[doc2.ID]: doc2
		},
		selectedDocId: null,
		isLoading: false,
		loaded: true,
		docTypes: [ mockDocType1, mockDocType2 ]
	};

	describe('LOAD_DOCS_SUCCESS', () => {
		type DocActions = typeof DocActions.loadDocsSuccess;
		function noExistingLoans(action: DocActions, docsInitialState: any, documents: Document[]) {
			const createAction = action({ documents });

			const result = fromDocs.docsReducer(docsInitialState, createAction);

			expect(result).toMatchSnapshot();
		}

		it('should add all loans in the payload when none exist', () => {
			noExistingLoans(DocActions.loadDocsSuccess, docsInitialState, [ doc1, doc2 ]);
		});
	});

	describe('LOAD_DOCTYPES_SUCCESS', () => {
		type DocActions = typeof DocActions.loadDocTypesSuccess;
		function noExistingLoanDetail(action: DocActions, docsInitialState, docTypes: CodeTable[]) {
			const createAction = action({ docTypes });

			const result = fromDocs.docsReducer(docsInitialState, createAction);

			expect(result).toMatchSnapshot();
		}

		it('should should return the full loan detail for requested loan number', () => {
			noExistingLoanDetail(DocActions.loadDocTypesSuccess, docsInitialState, [ docType1, docType2 ]);
		});
	});

});
