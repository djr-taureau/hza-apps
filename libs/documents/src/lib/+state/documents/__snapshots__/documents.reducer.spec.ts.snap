// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentsReducer LOAD_DOCS_SUCCESS should add all loans in the payload when none exist 1`] = `
Object {
  "docTypes": Array [
    Object {
      "CodeKey": "119",
      "CodeTable": "DocType",
      "CodeValue": "LM TAX RETURNS",
      "CodeValue1": "1",
      "CodeValue2": null,
      "CodeValue3": null,
      "CodeValue4": null,
      "CodeValue5": null,
      "CodeValue6": null,
      "CodeValue7": null,
      "CodeValue8": null,
      "CodeValue9": null,
      "LastUpdated": "2019-10-18T19:22:44.12",
      "LastUpdatedBy": "FAY\\\\colsen",
    },
    Object {
      "CodeKey": "12",
      "CodeTable": "DocType",
      "CodeValue": "BPO",
      "CodeValue1": "1",
      "CodeValue2": null,
      "CodeValue3": null,
      "CodeValue4": null,
      "CodeValue5": null,
      "CodeValue6": null,
      "CodeValue7": null,
      "CodeValue8": null,
      "CodeValue9": null,
      "LastUpdated": "2019-10-18T19:22:44.12",
      "LastUpdatedBy": "FAY\\\\colsen",
    },
  ],
  "entities": Object {
    "8004775": Object {
      "CreatedBy": "Fay_BulkLoad",
      "CreatedDate": "2016-08-15T14:39:56",
      "DeletedBy": null,
      "DeletedDate": null,
      "DocFileName": "TracyCascioNewAmCard",
      "DocType": "Miscellaneous",
      "DocumentID": "ad5e7986-5075-4601-8f75-6ea82d12f2e9",
      "Extension": ".pdf",
      "FileSize": null,
      "ID": 8004775,
      "IsDeleted": false,
      "Key1": null,
      "Key2": null,
      "Key3": null,
      "LoanNumber": "0000016094",
      "MeridianImport": true,
      "MetaDataType": null,
      "MetaDataValue": null,
      "ScanStatus": 1,
      "SourceID": 1,
      "UpdatedBy": null,
      "UpdatedDate": null,
    },
    "8004777": Object {
      "CreatedBy": "JNorton",
      "CreatedDate": "2013-03-05T11:46:36",
      "DeletedBy": null,
      "DeletedDate": null,
      "DocFileName": "16094 Romero Skip Trace 3-5-13",
      "DocType": "Miscellaneous",
      "DocumentID": "4e88c45a-0a52-47fc-aeba-fcd60c22607c",
      "Extension": ".pdf",
      "FileSize": null,
      "ID": 8004777,
      "IsDeleted": false,
      "Key1": null,
      "Key2": null,
      "Key3": null,
      "LoanNumber": "0000016094",
      "MeridianImport": true,
      "MetaDataType": null,
      "MetaDataValue": null,
      "ScanStatus": 1,
      "SourceID": 1,
      "UpdatedBy": null,
      "UpdatedDate": null,
    },
  },
  "ids": Array [
    8004775,
    8004777,
  ],
  "isLoading": false,
  "loaded": true,
  "selectedDocId": null,
}
`;

exports[`DocumentsReducer LOAD_DOCTYPES_SUCCESS should should return the full loan detail for requested loan number 1`] = `
Object {
  "docTypes": Array [
    Object {
      "CodeKey": "119",
      "CodeTable": "DocType",
      "CodeValue": "LM TAX RETURNS",
      "CodeValue1": "1",
      "CodeValue2": null,
      "CodeValue3": null,
      "CodeValue4": null,
      "CodeValue5": null,
      "CodeValue6": null,
      "CodeValue7": null,
      "CodeValue8": null,
      "CodeValue9": null,
      "LastUpdated": "2019-10-18T19:22:44.12",
      "LastUpdatedBy": "FAY\\\\colsen",
    },
    Object {
      "CodeKey": "12",
      "CodeTable": "DocType",
      "CodeValue": "BPO",
      "CodeValue1": "1",
      "CodeValue2": null,
      "CodeValue3": null,
      "CodeValue4": null,
      "CodeValue5": null,
      "CodeValue6": null,
      "CodeValue7": null,
      "CodeValue8": null,
      "CodeValue9": null,
      "LastUpdated": "2019-10-18T19:22:44.12",
      "LastUpdatedBy": "FAY\\\\colsen",
    },
  ],
  "entities": Object {
    "8004775": Object {
      "CreatedBy": "Fay_BulkLoad",
      "CreatedDate": "2016-08-15T14:39:56",
      "DeletedBy": null,
      "DeletedDate": null,
      "DocFileName": "TracyCascioNewAmCard",
      "DocType": "Miscellaneous",
      "DocumentID": "ad5e7986-5075-4601-8f75-6ea82d12f2e9",
      "Extension": ".pdf",
      "FileSize": null,
      "ID": 8004775,
      "IsDeleted": false,
      "Key1": null,
      "Key2": null,
      "Key3": null,
      "LoanNumber": "0000016094",
      "MeridianImport": true,
      "MetaDataType": null,
      "MetaDataValue": null,
      "ScanStatus": 1,
      "SourceID": 1,
      "UpdatedBy": null,
      "UpdatedDate": null,
    },
    "8004777": Object {
      "CreatedBy": "JNorton",
      "CreatedDate": "2013-03-05T11:46:36",
      "DeletedBy": null,
      "DeletedDate": null,
      "DocFileName": "16094 Romero Skip Trace 3-5-13",
      "DocType": "Miscellaneous",
      "DocumentID": "4e88c45a-0a52-47fc-aeba-fcd60c22607c",
      "Extension": ".pdf",
      "FileSize": null,
      "ID": 8004777,
      "IsDeleted": false,
      "Key1": null,
      "Key2": null,
      "Key3": null,
      "LoanNumber": "0000016094",
      "MeridianImport": true,
      "MetaDataType": null,
      "MetaDataValue": null,
      "ScanStatus": 1,
      "SourceID": 1,
      "UpdatedBy": null,
      "UpdatedDate": null,
    },
  },
  "ids": Array [
    8004775,
    8004777,
  ],
  "isLoading": false,
  "loaded": true,
  "selectedDocId": null,
}
`;
